### DISCUSSION TITLE
Is it possible to express that a result may be null/may not be null depending on a Boolean parameter?

### DISCUSSION THREAD

**Discussion by HugoRoss** [dotnet/csharplang#6194] [Q&A]
In C#, is it possible to express that a result may be null/may not be null depending on a Boolean parameter, something like the following (not compiling) code?

        [return: NotNullWhen(true, "noRowsThrows")]
        public DataRow? ExecuteToDataRow(SqlCommand command, bool noRowsThrows) {
            DataTable dataTable = ExecuteToDataTable(command);
            DataRowCollection rows = dataTable.Rows;
            switch (rows.Count) {
                case 1:
                    return rows[0];
                case < 1:
                    if (noRowsThrows) throw new ArgumentException($"The command did not return any rows!");
                    return null;
            }
            throw new ArgumentException($"The command returned multiple rows!");
        }

**Comment by RikkiGibson** [comment]
No. We suggest declaring multiple methods in a scenario like this.

If we introduced an attribute like this, the compiler would still only have a limited ability to track *what possible values does a boolean expression have*, and we would need to do significant work to improve that beyond simple scenarios like `ExecuteToDataRow(command, noRowsThrows: true)`.

```cs
    public DataRow? ExecuteToDataRow(SqlCommand command) {
        DataTable dataTable = ExecuteToDataTable(command);
        DataRowCollection rows = dataTable.Rows;
        switch (rows.Count) {
            case 1:
                return rows[0];
            case < 1:
                return null;
        }
        throw new ArgumentException($"The command returned multiple rows!");
    }


    public DataRow ExecuteToRequiredDataRow(SqlCommand command) {
        return ExecuteToDataRow(command) ?? throw new ArgumentException($"The command did not return any rows!");
    }
```
